servicetalk-http-api/src/main/java/io/servicetalk/http/api/AbstractZipContentCodec.java

```diff
@@ -0,0 +1,540 @@
+        final Buffer dst = allocator.newBuffer(ONE_KB);
+        InflaterInputStream input = null;
+        try {
+            input = newInflaterInputStream(asInputStream(src));
+
+            int read = dst.setBytesUntilEndStream(0, input, ONE_KB);
+            dst.writerIndex(read);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        } finally {
+            closeQuietly(input);
+        }
+
+        return dst;
+    }
+
+    @Override
+    public final Publisher<Buffer> decode(final Publisher<Buffer> from, final BufferAllocator allocator) {
+        return from.liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+
+            @Nullable
+            Buffer dst;
+            @Nullable
+            Inflater inflater;
+            @Nullable
+            ZLibStreamDecoder streamDecoder;
+
+            @Override
+            public void onSubscribe(final PublisherSource.Subscription subscription) {
+                dst = allocator.newBuffer(ONE_KB);
+                inflater = newRawInflater();
+                streamDecoder = new ZLibStreamDecoder(dst, inflater, supportsChecksum());
+                subscriber.onSubscribe(subscription);
+            }
+
+            @Override
+            public void onNext(@Nullable final Buffer src) {
+                assert streamDecoder != null;
+                assert src != null;
+
+                Buffer part;
+                try {
+                    if (streamDecoder.isFinished()) {
+                        throw new IllegalStateException("Stream encoder previously closed but more input arrived ");
+                    }
+
+                    part = streamDecoder.decode(src);
+                    subscriber.onNext(part != null ? part : EMPTY_BUFFER);
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                assert inflater != null;
+
+                inflater.end();
+                subscriber.onError(t);
+            }
+
+            @Override
+            public void onComplete() {
+                assert inflater != null;
+
+                inflater.end();
+                subscriber.onComplete();
+            }
+        });
+    }
+
+    private void closeQuietly(@Nullable final Closeable closeable) {
+        try {
+            if (closeable != null) {
+                closeable.close();
+            }
+        } catch (IOException e) {
+            LOGGER.error("Unexpected IO exception while closing buffer streams", e);
+        }
+    }
+
+    // Code forked from Netty's JdkZlibDecoder
+    static class ZLibStreamDecoder {
+        private static final int FHCRC = 0x02;
+        private static final int FEXTRA = 0x04;
+        private static final int FNAME = 0x08;
```

Comment:
This *is* Netty's implementation, unless you are referring to keeping everything on the transport layer, in which I did consider it, and chose to go with this approach for a few reasons.
- It allows from a much easier codec interface for a user to implement if they want custom encoder
- It allows for cold reads, on the receiving side, it will be decompressed when you need it, and not necessarily inside the event loop.
- It is consistent with gRPC experience

Reply:
Discussed offline the approach with `EmbeddedChannel`:
It won't require changing our public API or doing all the work in the transport layer on the event loop, we will just use a synthetic `EmbeddedChannel` internally instead of Input/Output streams to avoid manual parsing of the compression headers. The only downside is that we will need to depend on netty packages. Therefore, our implementations will need to be in a separate module, something like `servicetalk-content-coding-netty`.
Let's finish changes for the public API and then we can reconsider the internal implementations.